R/print.data.table.R

```diff
@@ -164,3 +186,47 @@ shouldPrint = function(x) {
 #   as opposed to printing a blank line, for excluding col.names per PR #1483
 cut_top = function(x) cat(capture.output(x)[-1L], sep = '\n')
 
+# to calculate widths of data.table for PR #4074
+nchar_width = function(x) {
+  each_width = nchar(x, type = "width")
+  widths = vector("numeric", ncol(x))
+  for (i in 1L:ncol(x)){
+    widths[i] = max(each_width[, i])
+  }
+  widths
+}
+# gets the width of the data.table at each column
+#   and compares it to the console width
+dt_width = function(x, class, row.names, col.names, names) {
+  widths = nchar_width(x)
+  if (class) widths = ifelse(widths < 6L, 6L, widths)
+  if (col.names == "none") names = sapply(names, nchar, type = "width") else names = 0L
+  dt_widths = ifelse(widths > names, widths, names)
+  rownum_width = if (row.names) max(nchar(as.character(rownames(x)), type = "width")) else 0L
```

Comment:
This one is also an expensive call we can avoid... if `x` has e.g. 1e9 rows, we would convert 1e9 integers to character, then run `nchar`. Whereas from `data.table(1:1e9)` we can see it's always printed as all digits (no scientific notation). So `if(row.names)`, `rownum_width` is `as.integer(ceiling(log10(nrow(x)))+1)`

Reply:
Could we grab the last row name assuming it will be the largest? e.g. `rownames(x)[length(x)]`? We would only need that one to add the widths.